---
title: "DEA"
author: "LuoYuting"
date: "2024-05-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:
#Count normalization
##Normalization
###Install packages
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DEA")

# install EdgeR package
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```

### Build a file description
```{r}
#file fileDesc
setwd('E:/UH/FOR-271 Practical Bioinformatics/topic 8/RNA_seq_DEA')
```

###read raw data
```{r}
#Set to the current work directory
Coinfection.targets<-read.delim("./data/fileDesc.txt")

#change the rawnames of the dataframe Coinfection.targets
rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
```

```{r}
#load package
library(edgeR)

#read files
Coinfection.orig <- readDGE(Coinfection.targets, header=F)

#check the dimension of the data set
dim(Coinfection.orig)

#chreck the first 6 rows of the data
head(Coinfection.orig)


```
```{r}
#extract counts dataframe
Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)

head(Coinfection.rawCount)

```
###Build meta data
```{r}
#define sampletype: Control and Ha-infection
sampletype <- factor(c(rep("Ha",3), rep("Ctr", 3)))

#Build meta data frame
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))

#check the column name of counts dataframe
colnames(Coinfection.orig$count)

```

```{r}
#check the rowname of meta dataframe
rownames(meta)
```
```{r}
#Check that sample names match in both files
all(colnames(Coinfection.orig$count) %in% rownames(meta))
```
###Create DESeqDataset object
```{r}
#load the package
library(DESeq2)
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)

#view dds
head(counts(dds))
```
###Generate size factors
```{r}
#generate size factors
dds <- estimateSizeFactors(dds)
#view
sizeFactors(dds)
```
###Generate the normalized counts
```{r}
normalized_counts <- counts(dds, normalized=TRUE)
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
```

#Sample-level quality Control
##Principal Component Analysis (PCA)
###Using our dataset for PCA
```{r}
#transform counts for data visualization
rld <- rlog(dds, blind=TRUE)

#plot PCA
plotPCA(rld, intgroup="sampletype")
```
```{r}
#save the plot
pdf("./results/PlotPCA_dds.pdf")
plotPCA(rld, intgroup="sampletype")
dev.off()
```
###Exercise:
####1.What does the above plot tell you about the similarity of samples?
The blue dots labeled 'Ha' are clustered together on the left side of the plot. The red dots labeled 'Ctr' are located on the right side of the plot. The samples within each group are similar to each other but are significantly different from the the other group.The blue dots are more similar to each other than the red dots. 
####2.Does it fit the expectation from the experimental design?
Yes, the plot fits the expectation from the experimental design. The'Ha' for Heterobasidion annosum infection and 'Ctr' for control are 
This is consistent with the experimental design, where the treatment is expected to induce changes in gene expression that are distinct from the control group.The distinct separation along PC1 also implies that the treatment effect is the most significant factor influencing gene expression variations in this dataset.
The clustering within each group suggests good experimental consistency and repeatability, each condition (control and treatment) behaved similarly across replicates.
####3.What do you think the %variance information (in the axes titles) tell you about the data in the context of the PCA?
PC1: 50% variance - This indicates that the first principal component accounts for 50% of the total variation in the dataset. This high percentage suggests that the major differences between the samples (most likely between the treated and control groups) are captured by PC1. It reflects a strong effect of the treatment on the gene expression profiles.
It shows the major factor that separates the samples in the dataset.
PC2: 25% variance - The second principal component accounts for 25% of the variance. This is also significant, indicating other variations in the dataset that are not captured by PC1 but are still relevant.
It shows the secondary but important differences among the samples.
The separation along PC1 between the two groups indicates a strong influence of the treatment on the gene expression profiles. Samples within each group cluster closely together, which suggests that the samples within each group (control or treatment) are more similar to each other than to samples from the other group. 
##Hierarchical Clustering Heatmap
```{r}
rld_mat <- assay(rld)
# compute the pairwise correlation values
rld_cor <- cor(rld_mat) 

#check the output
head(rld_cor)
```
```{r}
head(meta)
```
```{r}
#Install pheatmap package and load the package
install.packages("pheatmap")
?pheatmap # help page for pheatmap
```

```{r}
#plot the heatmap
library(pheatmap)
pheatmap(rld_cor, annotation = meta)
```
```{r}
pdf("./results/PlotHeatmap_dds.pdf")
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
        fontsize_row = 10, height=20)
```

#Differential expression analysis (DEA) using EdgeR
```{r}
#load EdgeR package
library(edgeR)
options(digits=3)
#read files
infection.targets<-read.delim("./data/fileDesc.txt")
#check Coinfection.targets
infection.targets
```
```{r}
#change the rawnames
rownames(infection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
#check the rawnames
infection.targets
```

```{r}
#read and merges the files
infection <- readDGE(infection.targets, header=F)
#Check the dimension of DGElist R object
dim(infection)

```
```{r}
head(infection)
```
```{r}
#Get the raw mapped count before filtering
infection.rawCount <- infection$count
head(infection.rawCount)

```
```{r}
library(ggplot2)
```

```{r}
#plot a histogram of the counts for a single sample, ‘Ha1’
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
```

```{r}
#Export the .png file
png("./results/count distribution.png", res=300, height=1800, width=1800)
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
dev.off()
```
```{r}
#Export raw count table into results folder
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv")
#Get the counts per million (TMM normalised) before filtering
infection.normCPM <- cpm(calcNormFactors(infection))
dim(infection.normCPM)
```
```{r}
head(infection.normCPM)
```
```{r}
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")
#Keep genes that are expressed at least 1 CPM in at least 3 libraries, normally it is the number of biological replicates of smaller group
infection.filtered <- rowSums(cpm(infection)>1) >=3
table(infection.filtered)
```
```{r}
#Libraries size of data BEFORE filtering
infection$samples$lib.size
```
```{r}
#cover the original file with our filter data
Infection <- infection[infection.filtered,]
#libraries size of data after filtering
colSums(Infection$counts)
```
```{r}
dim(Infection)
```
```{r}
#Update the filtered libraries size
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples

```
```{r}
#Performed normalisation with TMM method
Infection = calcNormFactors(Infection)
#The libraries after normalisation
Infection$samples
```
```{r}
#Get the counts per million (TMM normalised) after filtering
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))
#Export TMM normalized count table after filtering
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")
```
##differentailly gene expression analysis
```{r}
#Treatment factor
group<-factor(c('Ha','Ha','Ha',"Ctr","Ctr","Ctr"))
#Ctr is the intercept
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design
```
```{r}
#check outliers
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))
#Export the MDS plot to a PNG file
png("./results/plotMDS.Infection.png", res=300, height=1800, width=1800)
plotMDS(Infection, main="MDS plot of Infection RNA-Seq", labels=colnames(Infection$counts))
dev.off()
```
```{r}
#dispersion
Infection <- estimateGLMCommonDisp(Infection, Infection.design)
#Estimating the trended dispersion
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)
#Estimating the tagwwise dispersion
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)
#visual representation of the mean-variance relationship and BCV-mean logCPM relationship
plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
plotBCV(Infection)
```
```{r}
#Fit DGEList and design matrix to genewise GLM
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
```
```{r}
lrt.Ha_vs_Ctr <- glmLRT(Infection.fit, coef=2)  # to compare Ha vs Ctr (Ha_vs_Ctr)
t1<-topTags(lrt.Ha_vs_Ctr, n=nrow(Infection))
head(t1$table)

```
Total number of differentially expressed (DE) genes in different comparsion at FDR < 0.05
```{r}
summary(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))

```
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC > 0))
```
```{r}
lrt.Ha_vs_Ctr_UP <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC > 0)
#The DW-regulated genes (can change logFC to be more strict)
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC < 0))

```
```{r}
lrt.Ha_vs_Ctr_DW <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC < 0)
```
Differentially expressed transcripts’ tag
```{r}
DEtags.lrt.Ha_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))]
```
Export list of UP-regulated and DW-regulated transcripts
```{r}
write.csv(lrt.Ha_vs_Ctr_UP, file="./results/lrt.Ha_vs_Ctr_UP.csv")
write.csv(lrt.Ha_vs_Ctr_DW, file="./results/lrt.Ha_vs_Ctr_DW.csv")
```
To ask all the genes label as grey color
```{r}
Infection.colHavsCtr = rep('grey55', nrow(Infection))
```
To assign colour for DE transcripts
```{r}
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC <0 ] <- "blue"
```
Plot Smear plot with adjustment on Y-axis label
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, de.tags=DEtags.lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
```
Plot Smear plot with adjustment on Y-axis label and colour of DE tags changed
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="a infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
```
Plot Smear plot with adjustment on Y-axis label and colour of DE tags changed and export as png
```{r}
png("./results/plotSmear.InfectionRNAseq.png", res=300, height=1800, width=1800)
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
dev.off()
```






